online-shop-ui/src/context/shopping-cart-context.tsx

```diff
@@ -0,0 +1,71 @@
+import { createContext, ReactNode, useContext, useState } from "react";
+
+export type ShoppingCartProduct = {
+  id: string;
+  category: string;
+  productName: string;
+  price: number;
+  quantity: number;
+};
+
+type ShoppingCartProviderProps = {
+  children: ReactNode;
+};
+
+type ShoppingCartContextType = {
+  cartItems: ShoppingCartProduct[];
+  addItemToCart: (item: ShoppingCartProduct) => void;
+  clearCart: () => void;
+};
+
+export const ShoppingCartContext =
+  createContext<ShoppingCartContextType | null>(null);
+export const ShoppingCartAddContext = createContext<
+  (item: ShoppingCartProduct) => void
+>(() => {});
+
+export function ShoppingCartProvider({ children }: ShoppingCartProviderProps) {
+  const [shoppingCartItems, setShoppingCartItems] = useState<
+    ShoppingCartProduct[]
+  >([]);
+
+  function addItemToCart(item: ShoppingCartProduct) {
+    const isInCart: boolean =
+      shoppingCartItems.find((i) => i.id === item.id) !== undefined
+        ? true
+        : false;
+    if (isInCart) {
+      setShoppingCartItems((prevItems) =>
```

Comment:
You can write something like this:
```
setShoppingCartItems((prevItems) =>prevItems.map((i) =>
          i.id === item.id ? { ...i, quantity: i.quantity + 1 } : {...i, quantity: 1}
        )
      );
```

Reply:
If I only use this piece of code, after I add a product to the cart, it does not appear in it; I think the problem is that at first the cart is empty. I tried to solve the issue this way and it seems to work: 

setShoppingCartItems((prevItems) => {
      const existingItem = prevItems.find((i) => i.id === item.id);

      if (existingItem) {
        return prevItems.map((i) =>
          i.id === item.id ? { ...i, quantity: i.quantity + 1 } : i
        );
      } else {
        return [...prevItems, { ...item, quantity: 1 }];
      }
    });